x/pki/types/message_add_pki_revocation_distribution_point.go

```diff
@@ -0,0 +1,183 @@
+
+	if *msg.IsPAA {
+		if msg.Pid != 0 {
+			return pkitypes.NewErrNotEmptyPid("Product ID (pid) must be empty for root certificates when isPAA is true")
+		}
+
+		if !cert.IsSelfSigned() {
+			return pkitypes.NewErrRootCertificateIsNotSelfSigned(fmt.Sprintf("CRL Signer Certificate must be self-signed if isPAA is True"))
+		}
+
+		strVid, found := subjectAsMap["Mvid"]
+		if found {
+			vid, err := strconv.ParseInt(strings.Trim(strVid, "0x"), 16, 32)
+			if err != nil {
+				return err
+			}
+
+			if int32(vid) != msg.Vid {
+				return pkitypes.NewErrCRLSignerCertificateVidNotEqualMsgVid("CRL Signer Certificate vid must equal to message vid")
+			}
+		}
+	} else {
+		strPid, found := subjectAsMap["Mpid"]
+		if found {
+			pid, err := strconv.ParseInt(strings.Trim(strPid, "0x"), 16, 32)
+			if err != nil {
+				return err
+			}
+
+			if msg.Pid == 0 {
+				return pkitypes.NewErrPidNotFound("Product ID (pid) must be provided when it is found in non-root certificate")
+			}
+
+			if int32(pid) != msg.Pid {
+				return pkitypes.NewErrCRLSignerCertificatePidNotEqualMsgPid("CRL Signer Certificate pid must equal to message pid")
+			}
+		} else {
+			if msg.Pid != 0 {
+				return pkitypes.NewErrNotEmptyPid("Product ID (pid) must be empty when it is not found in non-root certificate")
+			}
+		}
+
+		if cert.IsSelfSigned() {
+			return pkitypes.NewErrNonRootCertificateSelfSigned(fmt.Sprintf("CRL Signer Certificate shall not be self-signed if isPAA is False"))
+		}
+
+		strVid, found := subjectAsMap["Mvid"]
+		if found {
+			vid, err := strconv.ParseInt(strings.Trim(strVid, "0x"), 16, 32)
+			if err != nil {
+				return err
+			}
+
+			if int32(vid) != msg.Vid {
+				return pkitypes.NewErrCRLSignerCertificateVidNotEqualMsgVid("CRL Signer Certificate vid must equal to message vid")
+			}
+		} else {
+			return pkitypes.NewErrVidNotFound("vid not found in CRL Signer Certificate subject")
+		}
+	}
+
+	if msg.DataFileSize == 0 && msg.DataDigest != "" {
+		return pkitypes.NewErrNonEmptyDataDigest("Data Digest must be provided only if Data File Size is provided")
+	}
+
+	if msg.DataFileSize != 0 && msg.DataDigest == "" {
+		return pkitypes.NewErrEmptyDataDigest("Data Digest must be provided if Data File Size is provided")
+	}
+
+	if msg.DataDigest == "" && msg.DataDigestType != 0 {
+		return pkitypes.NewErrNotEmptyDataDigestType("Data Digest Type must be provided only if Data Digest is provided")
+	}
+
+	if msg.DataDigest != "" && msg.DataDigestType == 0 {
+		return pkitypes.NewErrEmptyDataDigestType("Data Digest Type must be provided if Data Digest is provided")
+	}
+
+	if msg.RevocationType == allowedRevocationType && (msg.DataFileSize != 0 || msg.DataDigest != "" || msg.DataDigestType != 0) {
+		return pkitypes.NewErrDataFieldPresented("Data Digest, Data File Size and Data Digest Type must be omitted for Revocation Type 1")
+	}
+
+	match, _ := regexp.MatchString("^(?:[0-9A-F]{2})+$", msg.IssuerSubjectKeyID)
```

Comment:
It makes sense to move this into a function, as it's called in multiple places (Add, Delete, Update messages).